http-agent/src/main/java/org/motechproject/http/agent/listener/HttpClientEventListener.java

```diff
@@ -118,24 +122,79 @@ public void handle(MotechEvent motechEvent) {
         if (parameters.get(EventDataKeys.RETRY_INTERVAL) != null) {
             retryInterval = (long) parameters.get(EventDataKeys.RETRY_INTERVAL);
         }
-        ResponseEntity<?> retValue = null;
 
-        // defining a callable which attempts retry if response is not 2xx
+        return checkStatusCodeAndReturnResponse(url, requestData, method, basicRestTemplate, retryCount, retryInterval);
+
+    }
+
+    /**
+     * Handles an event and sends an http request with username and password parameters. Request sections such as headers, url or method are built from event
+     * parameters. Returns the response from the performed request. Retry count(default value is 1) and retry
+     * interval(default value is 0, expressed in milliseconds) can be specified by event parameters(keys:
+     * org.motechproject.http.agent.domain.EventDataKeys.RETRY_COUNT and org.motechproject.http.agent.domain.EventDataKeys.RETRY_INTERVAL).
+     *
+     * @param motechEvent the event which contains data for request
+     * @return response from the posted request
+     */
+    @MotechListener (subjects = { SendRequestConstants.SEND_REQUEST_SUBJECT })
+    public ResponseEntity<?> handleWithUserPasswordAndReturnType(MotechEvent motechEvent) {
+        Map<String, Object> parameters = motechEvent.getParameters();
+        final String url = String.valueOf(parameters.get(SendRequestConstants.URL));
+        Object requestData = parameters.get(SendRequestConstants.DATA);
+        String username = String.valueOf(parameters.get(SendRequestConstants.USERNAME));
+        String password = String.valueOf(parameters.get(SendRequestConstants.PASSWORD));
+        boolean redirection = (boolean) parameters.get(SendRequestConstants.REDIRECTION_ABILTY);
+        LOGGER.info(String.format("Posting Http request -- Url: %s, Data: %s",
+                url, String.valueOf(requestData)));
+
+        HttpEntity<Object> entity = new HttpEntity<>(requestData);
+
+        RestTemplate restTemplate;
+
+        if (StringUtils.isNotBlank(username)
+                || StringUtils.isNotBlank(password)) {
+            restTemplate = new RestTemplate(usernamePasswordRequestFactory(
+                    username, password));
+        } else {
+            restTemplate = basicRestTemplate;
+        }
+        
+        if (!redirection) {
```

Comment:
I really doubt that this if will help. We're not setting the follow redirect strategy here and the default behavior of the rest template is to avoid following redirects for POST request. Please check the following links for more information.

http://stackoverflow.com/questions/32392634/spring-resttemplate-redirect-302
http://stackoverflow.com/questions/29418583/follow-302-redirect-using-spring-resttemplate